packages/driver/src/cy/commands/sessions.ts

```diff
@@ -0,0 +1,822 @@
+          existingSession.hydrated = true
+
+          setActiveSession({ [existingSession.id]: existingSession })
+
+          dataLog.set({
+            consoleProps: () => getConsoleProps(existingSession),
+          })
+
+          // persist the session to the server. Only matters in openMode OR if there's a top navigation on a future test.
+          // eslint-disable-next-line no-console
+
+          // eslint-disable-next-line no-console
+          return Cypress.backend('save:session', { ...existingSession, setup: existingSession.setup.toString() }).catch(console.error)
+        })
+      }
+
+      // uses Cypress hackery to resolve `false` if validate() resolves/returns false or throws/fails a cypress command.
+      function validateSession (existingSession, _onFail) {
+        navigateAboutBlank()
+
+        const validatingLog = Cypress.log({
+          name: 'Validate Session',
+          message: '',
+          snapshot: false,
+          type: 'system',
+          state: 'passed',
+          event: true,
+          groupStart: true,
+        })
+
+        const onSuccess = () => {
+          validatingLog.set({
+            name: 'Validate Session: valid',
+            message: '',
+            type: 'system',
+            event: true,
+            state: 'warning',
+          })
+
+          Cypress.log({ groupEnd: true, emitOnly: true })
+        }
+
+        const onFail = (err) => {
+          _onFail(err, validatingLog)
+        }
+
+        let _commandToResume: any = null
+
+        let _didThrow = false
+
+        let returnVal
+
+        try {
+          returnVal = existingSession.validate()
+        } catch (e) {
+          onFail(e)
+
+          return
+        }
+
+        if (typeof returnVal === 'object' && typeof returnVal.catch === 'function' && typeof returnVal.then === 'function') {
+          return returnVal
+          .then((val) => {
+            if (val === false) {
+              // set current command to cy.session for more accurate codeFrame
+              cy.state('current', sessionCommand)
+              throw $errUtils.errByPath('sessions.callback_returned_false', { reason: 'resolved false' })
+            }
+
+            onSuccess()
+          })
+          .catch((err) => {
+            onFail(err)
+          })
+        }
+
+        cy.state('onCommandFailed', (err, queue, next) => {
+          const index = _.findIndex(queue.commands, (v: any) => _commandToResume && v.attributes.chainerId === _commandToResume.chainerId)
+
+          // attach codeframe and cleanse the stack trace since we may not hit the cy.fail callback
+          if (typeof err === 'string') {
+            err = new Error(err)
+          }
+
+          err.stack = $stackUtils.normalizedStack(err)
+
+          err = $errUtils.enhanceStack({
+            err,
+            userInvocationStack: $stackUtils.getUserInvocationStack(err, Cypress.state),
+            projectRoot: Cypress.config('projectRoot'),
+          })
```

Comment:
In what cases won't an error hit `cy.fail`? It's a bit concerning to see this here since all other failures go through the [`fail` function in `cy.js`](https://github.com/cypress-io/cypress/blob/8550842e3696095c9391e89f30c87af6369ecfb9/packages/driver/src/cypress/cy.js#L812) and it does more than is duplicated here.

Reply:
64485d1d updated comment